File: Stack/Design-Browser-History.js
class BrowserHistory {
  constructor(homepage) {
    this.history = [homepage];
    this.current = 0;
  }

  /**
   * @param {string} url
   * @return {void}
   */

  visit(url) {
    ++this.current;
    this.history[this.current] = url;
    this.history.length = this.current + 1;
  }

  /**
   * @param {number} steps
   * @return {string}
   */
  back(steps) {
    if (steps > this.current) {
      this.current = 0;
      return this.history[this.current];
    }
    this.current -= steps;
    return this.history[this.current];
  }

  /**
   * @param {number} steps
   * @return {string}
   */
  forward(steps) {
    if (this.current + steps > this.history.length - 1) {
      this.current = this.history.length - 1;
      return this.history[this.current];
    }

    this.current += steps;
    return this.history[this.current];
  }
}

File: Rust/Design-Browser-History.rs
struct BrowserHistory {
    history: Vec<String>,
    current: usize,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl BrowserHistory {
    fn new(homepage: String) -> Self {
        Self {
            history: vec![homepage],
            current: 0,
        }
    }

    fn visit(&mut self, url: String) {
        self.current += 1;
        self.history.splice(self.current.., std::iter::once(url));
    }

    fn back(&mut self, steps: i32) -> String {
        self.current = self.current.saturating_sub(steps as usize);
        self.history[self.current].clone()
    }

    fn forward(&mut self, steps: i32) -> String {
        self.current = (self.current + steps as usize).min(self.history.len() - 1);
        self.history[self.current].clone()
    }
}

fn main() {}
